lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -249,19 +349,32 @@ swiftModuleIsUpToDate(clang::vfs::FileSystem &FS,
   assert(VI.name == ModuleID.first.str() &&
          "we built a module at this path with a different name?");
 
-  for (auto In : AllDeps) {
-    if (OuterTracker)
-      OuterTracker->addDependency(In.Path, /*IsSystem=*/false);
-    if (!dependencyIsUpToDate(FS, In, OutPath, Diags, ModuleID.second)) {
-      LLVM_DEBUG(llvm::dbgs() << "Dep " << In.Path
-                 << " is directly out of date\n");
-      return false;
-    }
-    LLVM_DEBUG(llvm::dbgs() << "Dep " << In.Path << " is up to date\n");
-  }
-  return true;
+  return dependenciesAreUpToDate(AllDeps, FS, OutPath, Diags,
+                                 ModuleID.second, OuterTracker);
 }
 
+// Check that a "forwarding" .swiftmodule file is at least as new as all the
+// dependencies it read when it was built last time.
+static bool forwardingModuleIsUpToDate(ModuleDiscovery &module,
+  clang::vfs::FileSystem &fs, DiagnosticEngine &diags,
+  SourceLoc diagLoc, DependencyTracker *tracker,
+  SmallVectorImpl<FileDependency> &deps) {
+
+  module.loadForwardingModule(fs);
+  auto &fwd = module.getForwardingModule();
+
+  for (auto dep : fwd.dependencies) {
```

Comment:
Document that this is expensive, and should be replaced once we wrap `clang::FileManager`. That'll be in another commit.